File: ui-preview/src/main/java/l/files/ui/preview/MemCache.java
package l.files.ui.preview;

import android.support.v4.util.LruCache;

import l.files.fs.Path;
import l.files.fs.Stat;
import l.files.ui.base.graphics.Rect;

import static java.util.concurrent.TimeUnit.MILLISECONDS;

abstract class MemCache<K, V> extends Cache<V> {

    private long lastModifiedTime(Stat stat) {
        return stat.lastModifiedTime().to(MILLISECONDS);
    }

    @Override
    V get(Path path, Stat stat, Rect constraint, boolean matchTime) {
        K key = getKey(path, constraint);
        Snapshot<V> value = delegate().get(key);
        if (value == null) {
            return null;
        }
        if (matchTime && lastModifiedTime(stat) != value.time()) {
            return null;
        }
        return value.get();
    }

    abstract K getKey(Path path, Rect constraint);

    @Override
    Snapshot<V> put(Path path, Stat stat, Rect constraint, V value) {
        return delegate().put(
                getKey(path, constraint),
                Snapshot.of(value, lastModifiedTime(stat)));
    }

    Snapshot<V> remove(Path path, Rect constraint) {
        return delegate().remove(getKey(path, constraint));
    }

    abstract LruCache<K, Snapshot<V>> delegate();

}

File: ui-preview/src/main/java/l/files/ui/preview/MemCache.kt
package l.files.ui.preview

import androidx.collection.LruCache
import l.files.ui.base.graphics.Rect
import java.nio.file.Path
import java.nio.file.attribute.BasicFileAttributes
import java.time.Instant

internal abstract class MemCache<K, V> : Cache<V> {

  private fun lastModifiedTime(attrs: BasicFileAttributes): Long =
    attrs.lastModifiedTime().toMillis()

  override operator fun get(
    path: Path,
    time: Instant,
    constraint: Rect,
    matchTime: Boolean
  ): V? {
    val key = getKey(path, constraint)
    val value = delegate[key] ?: return null
    return if (matchTime && time != value.time) {
      null
    } else {
      value.value
    }
  }

  abstract fun getKey(path: Path, constraint: Rect): K

  override fun put(
    path: Path,
    time: Instant,
    constraint: Rect,
    value: V
  ): Snapshot<V>? = delegate.put(
    getKey(path, constraint),
    Snapshot(value, time)
  )

  fun remove(path: Path, constraint: Rect): Snapshot<V>? =
    delegate.remove(getKey(path, constraint))

  abstract val delegate: LruCache<K, Snapshot<V>>
}
